entity_api/src/note.rs

```diff
@@ -79,7 +104,27 @@ pub async fn find_by(
         }
     }
 
-    Ok(query.all(db).await?)
+    match query.all(db).await {
```

Comment:
Disclaimer:  This is from own standards in my brain which I never put on paper so take it with a grain of salt.

It was an intentional decision to return the empty collection when a query that results in a `WHERE`  SQL clause returns 0 records. This is conventional in other ORMs that I've used. 

`RecordNotFound` should only be returned for queries that result in a `SELECT` SQL statement.

The reason being is exactly so that we didn't have to do the extra check like `!notes.is_empty()`

I'm curious what the reasoning here is and if we could provide the same benefits without the extra nesting and conditional logic.

Reply:
@calebbourg You wouldn't ever get an error for the empty case, so you'd always get a 200 OK which is not correct, the record isn't found. So that second match branch, i.e. `Err(err)` does not occur when a `Note` isn't found, for example, if you provide an id of a Note that doesn't exist.

Reply:
I'm proposing that we do actually want to return a `200` when no records exist for index fetches to align with the [JSON API convention](https://jsonapi.org/format/#fetching-resources-responses-200).

> server MUST respond to a successful request to fetch a resource collection with an array of [resource objects](https://jsonapi.org/format/#document-resource-objects) or an empty array ([]) as the response document’s primary data.

I think the philosophy is that the request itself is still "successful" in that there were no exceptions or auth issues. We just happened to return 0 records.

Reading more into the spec I think that the [204 No Content](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/204) we're mapping to `RecordNotFound` is not semantically inline with the HTTP standard as it states.

> This might be used, for example, when implementing "save and continue editing" functionality for a wiki site. In this case a [PUT](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/PUT) request would be used to save the page, and the 204 No Content response would be sent to indicate that the editor should not be replaced by some other page.


Which I read as "something was successful here but this endpoint itself doesn't return anything meaningful". The fact that adhering to the standard yields less matching and conditional logic in the code leads me to believe that seaORM may have designed their API to support these patterns. 

This is all interpretation, of course, and I'm learning these standards as we go.

Reply:
Now having read the JSON API spec in this area, here's what I see it saying:



https://jsonapi.org/format/#fetching-resources-responses-404

There are 3 cases:
1) Return `[]` for a resource that does exist, e.g. http://localhost/articles/ but there are no articles yet but also the URI exists.
2) Return a `null` for an nested entity that doesn’t exist (e.g. an article that has an author that hasn’t been defined yet) under the requested resource
3) A request for an instance of an entity at a specific URI that doesn’t exist should return a 404